tests/proptest.rs

```diff
@@ -0,0 +1,441 @@
+            for _ in 0..3 {
+                match (&one, &resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver)) {
+                    (Ok(l), Ok(r)) => assert_eq!(l, r),
+                    (Err(PubGrubError::NoSolution(derivation_l)), Err(PubGrubError::NoSolution(derivation_r))) => {
+                        prop_assert_eq!(
+                            format!("{}", DefaultStringReporter::report(&derivation_l)),
+                            format!("{}", DefaultStringReporter::report(&derivation_r))
+                        )},
+                    _ => panic!("not the same result")
+                }
+            }
+        }
+    }
+
+    #[test]
+    /// [ReversedDependencyProvider] change what order the candidates
+    /// are tried but not the existence of a solution
+    fn prop_reversed_version_errors_the_same(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666)
+    )  {
+        let minimal_provider = ReversedDependencyProvider(dependency_provider.clone());
+        for (name, ver) in cases {
+            let l = resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver);
+            let r = resolve(&TimeoutDependencyProvider::new(minimal_provider.clone(), 50_000), name, ver);
+            match (&l, &r) {
+                (Ok(_), Ok(_)) => (),
+                (Err(_), Err(_)) => (),
+                _ => panic!("not the same result")
+            }
+        }
+    }
+
+    #[test]
+    fn prop_removing_a_dep_cant_break(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666),
+        indexes_to_remove in prop::collection::vec((any::<prop::sample::Index>(), any::<prop::sample::Index>(), any::<prop::sample::Index>()), 1..10)
+    )  {
+        let packages: Vec<_> = dependency_provider.packages().collect();
+        let mut removed_provider = dependency_provider.clone();
+        for (package_idx, version_idx, dep_idx) in indexes_to_remove {
+            let package = package_idx.get(&packages);
+            let versions = dependency_provider.list_available_versions(package).unwrap();
+            let version = version_idx.get(&versions);
+            let dependencys: Vec<_> = dependency_provider.get_dependencies(package, version).unwrap().unwrap().into_iter().collect();
+            if !dependencys.is_empty() {
+                let dependency = dep_idx.get(&dependencys).0.clone();
+                removed_provider.add_dependencies(
+                    **package,
+                    *version,
+                    dependencys.into_iter().filter(|x| x.0 != dependency)
+                )
+            }
+        }
+        for (name, ver) in cases {
+            if resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver).is_ok() {
+                prop_assert!(
+                    resolve(&TimeoutDependencyProvider::new(removed_provider.clone(), 50_000), name, ver).is_ok(),
```

Comment:
I generally don't want to `resolve` on a generated `dependency_provider` without some kind of timeout. It is frustrating when CI fails with "terminated do to no output in 10 min" with no info on what input is slow. (looks like that limitation may have been from my Travis days, Github Actions may not have that limitation, but a hang is also frustrating.

What behaviour do you think the test should have when one of these hangs?

Reply:
I think it's probable that those two won't take the same amount of time and it's not what this test tries to ensure, so if at least one of the two `resolve` timeouts the test should pass in my opinion. Passing or not passing that test will be much more correlated to the shape of index than to its size if there is an issue somewhere. In fact it would make sense to have different registry strategies, of different sizes and shapes, but that can be explored later.

Concretely, we can let it fail for now. Gather the examples when it fails to rerun them locally and figure if those are true failures. When a few of these happen we can see if it's worth continuing to fail or not.

Reply:
We will definitely have to work on all of these tests, to ensure they maintain a good signal to noise ratio. If they turn out not to pull their weight, we can alter or remove!
Just to clarify are you ok with merging as is?